feat(robot-server): Add `PATCH /errorRecovery/settings` endpoint with `enableErrorRecovery` setting

SyntaxColoring opened the pull request:
## Overview

Another step towards EXEC-719.

## Test Plan and Hands on Testing

* [x] Play around with the endpoints with a dev server and make sure they work as expected, including that the value persists across boots.

## Changelog

This adds a `disableErrorRecovery` entry ~~to the `/settings` endpoints.~~ The value persists across boots via prior work in #16333. It does not actually affect robot behavior yet; follow-up work will have to connect it to `RunOrchestrator` etc.

Historically, the `/settings` endpoints have been a dumb window to `opentrons.config.advanced_settings`. We're skeptical of continuing that pattern, because: (1) sometimes these settings actually have nothing to do with the `opentrons` package; and (2) the `opentrons` package is a library that's installed on users' computers, and we don't want it storing global config that only makes sense when running on an actual OT-2 or Flex.

So, this PR goes to a little bit of pain to define the the new setting in `robot_server`, not in `opentrons`. It's kind of half-assed, sorry.

## Review requests

~~I'm unsure about my overall API design. Are the `/settings` endpoints the right place to add this?~~

~~Right now, the `/settings` endpoints are "dynamic": you can only know what options are available to change by running the server and asking it via `GET /settings`. Or by inspecting the Python source. The list of settings isn't statically available in the OpenAPI spec or anything.~~

~~This works well when we expect the client to dumbly render whatever the server returns. It's good for stuff like a feature flags page. But it makes less sense for settings that are actual *settings*—actual *product features*—where we want UX designs to dictate specific placement and copy.~~

~~So I guess I'm unsure of whether we want to keep extending `/settings` with stuff like this.~~

## Risk assessment

Low.

SyntaxColoring committed:
Delete unused SettingException class.

SyntaxColoring committed:
Wire up to HTTP API.

SyntaxColoring committed:
Todo comments in advanced_settings.py.

SyntaxColoring committed:
Light cleanup to set_oem_mode_request().

Add type annotations, make it private, and fix a type safety thing.

SyntaxColoring committed:
Add FastAPI dependency.

sfoster1 reviewed (changes requested):
I agree with your cover comment - I do not think this should be in `/settings`, which regardless of how it started is really now just for feature flags. i would support either having a new route or just having like `/settings/fixed` ("fixed" as opposed to "dynamic", not "fixed" as opposed to "broken") or something

vegano1 commented on robot-server/robot_server/service/legacy/routers/settings.py:
```diff
@@ -121,38 +103,118 @@ async def post_settings(
+
+    # Settings defined by the `opentrons` package:
+    else:
+        # Note: we're storing the new value of enableOEMMode via
+        # `advanced_settings.set_adv_setting()` to preserve prior behavior just in
+        # case, but I'm pretty sure that that has no functional effect. The actual
+        # meaningful storage happens via `set_oem_mode_request()`. We should remove the
+        # `set_adv_setting()` call for this when we confirm that it's safe.
+        if update.id == "enableOEMMode" and robot_type == RobotTypeEnum.FLEX:
```
Unfortunately this is not accurate, the client uses the `enableOEMMode` advanced setting for the following

- The desktop app uses `enableOEMMode` to determine if the OEM mode toggle should be set or not.
- The ODD uses `enableOEMMode` to render in `branded` or `unbranded` mode

Right now without the `enableOEMMode` advanced setting the OEM Mode feature would not work

vegano1 reviewed (commented):
Left a comment for oem mode

SyntaxColoring committed:
Move store to error_recovery.settings.

SyntaxColoring committed:
Implement `GET/PATCH/DELETE /errorRecovery/settings`.

SyntaxColoring committed:
Add Tavern tests.

SyntaxColoring committed:
Revert "Wire up to HTTP API."

This reverts commit 52c648228040399997947d4710585646df236640.

SyntaxColoring committed:
Revert "Todo comments in advanced_settings.py."

This reverts commit b6e802b6a55ac36fa8bdfce01c124f75655d67a0.

SyntaxColoring committed:
Revert "Delete unused SettingException class."

This reverts commit 26762b27a10b5f665796dc5b237accc2cbebe481.

SyntaxColoring committed:
Fix no-op patch effectively setting enabled=false.

SyntaxColoring committed:
Describe default in field documentation.

SyntaxColoring committed:
Once again flip "disable error recovery" back to to "enable error recovery."

This reverts commit 8b33cade81f9d3bbf1efc3aa2ed6bdf6c158bb6f.

As described in that commit message, "disable" was kind of our only choice when this was part of the /settings endpoints. But now we're not using /settings, so we should follow the general rule of thumb of phrasing boolean settings in the affirmative.

SyntaxColoring commented on robot-server/tests/integration/http_api/error_recovery/test_settings.tavern.yaml:
I'm once again omitting router-level tests with Decoy mocks, because this is a trivial CRUD endpoint and they would be wholly redundant with this Tavern test.

SyntaxColoring reviewed (commented):
Moved from the existing `/settings` (feature flags) endpoint to a new `PATCH /errorRecovery/settings` endpoint.

I didn't feel good about trying to name and design a more general replacement for `/settings` ("`/settings/fixed`") right now.

SyntaxColoring committed:
Revert unnecessary _set_oem_mode_request() movement.

sfoster1 reviewed (approved):
Yeah, this looks good to me. Agree with the choice of route, too

TamarZanzouri reviewed (approved):
Nice job!!!
